soda/contracts/README.md

```diff
@@ -0,0 +1,107 @@
+# Goal and purpose of data contracts in Soda
+
+Soda's goal is to be the best enforcement engine for data contract implementations. For
+this purpose we have created a new YAML format that aligns with data contracts from a producer
+perspective.  This new format enables a very easy translation from any data contract
+YAML language to Soda's enforcable contract language.
+
+# Example
+
+```yaml
+name: DIM_CUSTOMER
+
+columns:
+  - name: id
+    data_type: character varying
+    unique: true
+  - name: cst_size
+    data_type: decimal
+  - name: cst_size_txt
+    valid_values: [1, 2, 3]
+  - name: distance
+    data_type: integer
+  - name: country
+    data_type: varchar
+    not_null: true
+  - name: ts
+
+checks:
+  - avg(distance) between 400 and 500
+```
+
+# Top level keys
+
+| Key | Description | Required | YAML data type |
+| --- | ----------- | -------- | -------------- |
+| `name` | Name of the dataset as in the SQL engine. | Required | string |
+| `columns` | Schema specified as a list of columns.  See below for the column format | Required| list of objects |
+| `checks` | SodaCL checks.  See [SodaCL docs](https://docs.soda.io/soda-cl/metrics-and-checks.html) | Optional | list of SodaCL checks |
+
+# Column keys
+
+| Key | Description | Required | YAML data type |
+| --- | ----------- | -------- | -------------- |
+| `name` | Name of the column as in the SQL engine. | Required | string |
+| `data_type` | Ensures verification of this physical data type as part of the schema check. Must be the name of the data type as in the SQL engine | Optional | string |
+| `not_null` | `not_null: true` Ensures a missing values check | Optional | boolean |
+| `missing_*` | Ensures a missing values check and uses all the `missing_*` configurations | Optional | list of strings or numbers |
+| `valid_*` * `invalid_*` | Ensures a validity check and uses all the `valid_*` and `invalid_*` configurations | Optional | list of strings or numbers |
+| `unique` | `unique: true` ensures a uniqueness check | Optional | boolean |
+
+# Contract to SodaCL translation
+
+## The schema check
+
+A contract verification will always check the schema.  The contract schema check will verify that the list of columns
+in the database dataset matches *exactly* with the columns in the contract file.  All columns listed in the contract
```

Comment:
so this basically leads to a schema check that warns of `any` changes happen correct? In fact, maybe if you can show me what CL the example above would lead to that would be even easier to talk about it (as I won't review the code -- I leave that to Milan)

Which would also mean that the index (order) of the columns is also checked? If so, I don't know if I really like that. I find that more often than not, index (column position) changes are not an issue, however, they happen very often.

If the answer to the above question was going to be yes, then I was wondering if we could somewhere let people tell us what kind of schema check they want (or at least how strict). But let's see what the answer to the question is.

Reply:
> a schema check that warns of any changes happen correct?

Warns if there is any difference between the contract schema and the measured schema of the table.  (It is not related change over time)

> Which would also mean that the index (order) of the columns is also checked?

At the moment, the index is not checked.  I agree that this is too volatile and not a common cause of issues.

Reply:
> Warns if there is any difference between the contract schema and the measured schema of the table. (It is not related change over time)

Ah you're not leveraging the schema check we have in core I see. Ok I guess then all my questions go away :)